File: cpp/String/ReverseWordsString.cpp
#include<bits/stdc++.h>
using namespace std;

string reverseWords(string s) {
        // stack<string>st;
        // string str="";
        // for(int i=0;i<s.length();i++)
        // {
        //     if(s[i]==' ')
        //     {
        //         if(!str.empty())
        //         {   
        //             st.push(str);
        //             str="";
        //         }
        //     }
        //     else
        //     {
        //         str+=s[i];
        //     }
        // }
        // if(!str.empty())
        // {
        //     st.push(str);
        // }
        // string ans="";
        // while(!st.empty())
        // {
        //     ans+=st.top();
        //     st.pop();
        //     if (!st.empty()) {
        //         ans += " ";
        //     }
        // }
        // return ans;

        int left=0;
        int right=s.length()-1;

        while (left <= right && s[left] == ' ') left++;
        while (right >= left && s[right] == ' ') right--;

        string temp="";
        string ans="";

        while(left<=right)
        {
            char ch=s[left];
            if(ch!=' ')
            {
                temp+=ch;
            }
            else if(ch==' ' && temp!="")
            {
                if(ans!="") ans=temp+" "+ans;
                else ans=temp;
                temp="";
            }
            left++;
        }
        if(temp!=""){
            if(ans!="") ans=temp+" "+ans;
            else ans=temp;
        }
        return ans;
    }


int main()
{
    string s;
    cin>>s; 
    cout<<reverseWords(s);
    return 0;
}

//TC=O(N) SC=O(N)
//TC=O(N) SC=O(1)

File: java/String/ReverseWordsString.java
package String;
import java.util.*;
public class ReverseWordsString {

    public static String reverseWords(String s) {
        // String ans = "";
        // String word = "";
        // for (int i = 0; i < s.length(); i++) {
        //     if (s.charAt(i) == ' ') {
        //         ans = word + " " + ans;
        //         word = "";
        //     } else {
        //         word = word+s.charAt(i);
        //     }
        // }
        // return word + " " + ans;

        Stack<String> st = new Stack<String>();
        String str="";
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)==' '){
                if(!str.isEmpty()){
                    st.push(str);
                    str="";
                }
            }
            else{
                str+=s.charAt(i);
            }
        }
        if(!str.isEmpty()){
            st.push(str);
        }
        String ans="";
        while(st.size()!=1){
            ans+=st.pop()+" ";
        }
        return ans+st.pop();
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String s = sc.nextLine();
        String ans = reverseWords(s);
        System.out.println(ans);
        sc.close();
    }
}
